sw/device/silicon_creator/lib/drivers/spi_device.c

```diff
@@ -0,0 +1,66 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/silicon_creator/lib/drivers/spi_device.h"
+
+#include "sw/device/lib/base/abs_mmio.h"
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/silicon_creator/lib/error.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "spi_device_regs.h"  // Generated.
+
+enum {
+  /**
+   * Base address of the spi_device registers.
+   */
+  kBase = TOP_EARLGREY_SPI_DEVICE_BASE_ADDR,
+};
+
+void spi_device_init(void) {
+  // CPOL = 0, CPHA = 0, MSb-first TX and RX, 3-byte addressing.
+  uint32_t reg = bitfield_bit32_write(0, SPI_DEVICE_CFG_CPOL_BIT, false);
+  reg = bitfield_bit32_write(reg, SPI_DEVICE_CFG_CPHA_BIT, false);
+  reg = bitfield_bit32_write(reg, SPI_DEVICE_CFG_TX_ORDER_BIT, false);
+  reg = bitfield_bit32_write(reg, SPI_DEVICE_CFG_RX_ORDER_BIT, false);
+  reg = bitfield_field32_write(reg, SPI_DEVICE_CFG_TIMER_V_FIELD, 0x7f);
+  reg = bitfield_bit32_write(reg, SPI_DEVICE_CFG_ADDR_4B_EN_BIT, false);
+  reg = bitfield_bit32_write(reg, SPI_DEVICE_CFG_MAILBOX_EN_BIT, false);
+  abs_mmio_write32(kBase + SPI_DEVICE_CFG_REG_OFFSET, reg);
+
+  // JEDEC manufacturer and device ID.
+  // spi_device sends these in the following order: continuation codes,
+  // manufacturer ID, LSB of device ID, and MSB of device ID.
+  reg = bitfield_field32_write(0, SPI_DEVICE_JEDEC_CC_CC_FIELD,
+                               kSpiDeviceJedecContCode);
+  reg = bitfield_field32_write(reg, SPI_DEVICE_JEDEC_CC_NUM_CC_FIELD,
+                               kSpiDeviceJedecContCodeCount);
```

Comment:
Note: we'll have to be careful about this.  Some "client" implementations assume the JEDEC ID will always be 3 bytes and can't handle continuation code.

I don't think we should change this _yet_, but be aware that we may have to change this in the future.

Reply:
Ack. Just an FYI for @eunchan (because I'm not sure what the desired behavior is): I've just tried setting `kSpiDeviceJedecContCodeCount` to 0, first 3 bytes are as expected but the rest is 0x0 (if SCK keeps toggling).

Reply:
So, what value should it be? I can quickly fix it if it should be `0xFF`.

Reply:
I'm not sure if it's "don't care" or should be a particular value. @cfrantz, @a-will any thoughts?

Reply:
I suspect it's "don't care". Some parts actually tack on customer-programmed values to the end, but not all do.

Reply:
My experience with the H1 chip is that if you clocked in more bytes than the JEDEC ID, it would simply repeat the JEDEC ID pattern over and over.  It's probably OK to just emit 0x00 or 0xFF for excess bytes.

Reply:
@alphan Just to clarify, the number of continuation bytes can't depend on the opcode, so it won't help to change the hardware here. This will be a decision about which response to give to the 0x9F command. If we wanted to support such non-compliant hosts, we'd have to give a non-compliant response ourselves (i.e. use some other company's JEDEC ID).

If we want it to be an option which we use, I guess you could use pin strapping or fuse or similar.

Reply:
Thanks @a-will, yes this would be non-compliant and my suggestion probably doesn't make a whole lot of sense since I was thinking about using another command for this non-compliant response _assuming that_ we can teach this to such hosts.